hers_diagnostic_output/hers_diagnostic_output.py

```diff
@@ -292,213 +1043,228 @@ def multiply_energy_use_and_emission_factors(self):
                         "lb/kBtu",
                     )
                 else:
-                    self.emissions[home_type] += (
+                    total_emissions += (
                         self.data_cache[(fuel_type, home_type, "annual")]
                         * self.fuel_emission_factors[fuel_type]
                     )
+        return total_emissions
 
-    def calculate_annual_hourly_co2_emissions(self):
+    def calculate_annual_hourly_co2_emissions(self, home_type):
         # retrieve energy use for each subsystem and multiply energy by emissions factors
 
-        for home_type in ["rated_home", "co2_reference_home"]:
-            for system_type in self.data[f"{home_type}_output"]:
-                if system_type in self.system_types_system_output:
-                    for system_index in range(
-                        len(self.data[f"{home_type}_output"][system_type])
-                    ):
-                        for energy_use in self.data[f"{home_type}_output"][system_type][
-                            system_index
-                        ]["energy_use"]:
-                            self.calculate_energy_type_total_energy(
-                                energy_use, home_type
-                            )
-                elif system_type in self.other_end_uses_energy:
-                    for energy_use in self.data[f"{home_type}_output"][system_type]:
+        for end_use in self.data[f"{home_type}_output"]:
```

Comment:
As it stands, `data_cache` is only ever used within this method (and the methods it calls). I don't think it is really being used to store calculated values to avoid re-calculation. I think we should instead have separate methods to "get" data from the cache. For example`:

```python
def get_annual_energy(self, home_type, end_use, fuel_type):
    if (home_type, end_use, fuel_type) not in self.annual_energy_cache:
        self.annual_energy_cache[(home_type, end_use, fuel_type)] = ... # Calculate this combination
    return self.annual_energy_cache[(home_type, end_use, fuel_type)]
```

You can then make higher level functions to aggregate these higher. For example:

```python
def get_annual_fuel_energy(self, home_type, fuel_type):
    if (home_type, fuel_type) not in self.annual_fuel_energy_cache:
        self.annual_fuel_energy_cache[(home_type, fuel_type)] = 0.0
        for end_use in end_uses:
            self.annual_fuel_energy_cache[(home_type, fuel_type)] += self.annual_energy_cache[(home_type, end_use, fuel_type)]
    return self.annual_fuel_energy_cache[(home_type, fuel_type)]
```

You can then have something similar for hourly values, but since the hourly values are only used in the CO2 Index calculation, you probably don't need to build a cache for it.